proposals/NNNN-global-actor-isolated-types-usability.md

```diff
@@ -0,0 +1,267 @@
+
+Next, under the current concurrency rules, globally isolated functions and closures do not implicitly conform to `Sendable`. This impacts usability, because these closures cannot themselves be captured by `@Sendable` closures, which makes them unusable with `Task`:
+
+```swift 
+func test() {
+  let closure: @MainActor () -> Void = {
+    print("hmmmm")
+  }
+
+  Task {
+    // error: capture of 'closure' with non-sendable type '@MainActor () -> Void' in a `@Sendable` closure
+    await closure()
+  }
+}
+```
+
+In the above code, the closure is global-actor-isolated, so it cannot be called concurrently. The compiler should be able to infer the `@Sendable` attribute. Because of the same reason, globally isolated closures should be allowed to capture non-`Sendable` values.
+
+
+Finally, the current diagnostic for a global-actor-isolated subclass of a non-isolated superclass is too restrictive:
+
+```swift
+class NotSendable {}
+
+
+@MainActor
+class Subclass: NotSendable {} // error: main actor-isolated class 'Subclass' has different actor isolation from nonisolated superclass 'NotSendable'
+```
+
+Because global actor isolation on a class implies a `Sendable` conformance, adding isolation to a subclass of a non-`Sendable` superclass can circumvent `Sendable` checking:
+
+```swift
+func computeCount() async -> Int { ... }
+
+class NotSendable {
+  var mutableState = 0
+  func mutate() async {
+    let count = await computeCount()
+    mutableState += count
+  }
+}
+
+@MainActor
+class Subclass: NotSendable {}
+
+func test() async {
+  let c = Subclass()
+  await withDiscardingTaskGroup { group in
+    group.addTask {
+      await c.mutate()
+    }
+
+    group.addTask { @MainActor in
+      await c.mutate()
+    }
+  }
+}
+```
+
+In the above code, an instance of `Subclass` can be passed across isolation boundaries because `@MainActor` implies that the type is `Sendable`. However, `Subclass` inherits non-isolated, mutable state from the superclass, so this `Sendable` conformance allows smuggling unprotected shared mutable state across isolation boundaries to create potential for concurrent access. For this reason, the warning about adding isolation to a subclass was added in Swift 5.10, but this restriction could be lifted by instead preventing the subclass from being `Sendable`.
+
+## Proposed solution
+
+We propose that:
+
+- `Sendable` properties of a global-actor-isolated value type would be treated `nonisolated` as inferred within the module.
+- `@Sendable` would be inferred for global-actor-isolated functions and closures. Additionally, globally isolated closures would be allowed to capture non-`Sendable` values.
+- The programmer would be able to suppress the automatic conformance inferred via the above rule using the new `@~Sendable` attribute. By analogy, introduce a new `~Sendable` protocol to indicate that a nominal type is not `Sendable`.
```

Comment:
I'm not sure this carries its weight.  Do we have any use for this?

Reply:
I have discussed this with Holly and we also feel like this change can be omitted because there's no obvious use case. We can note this in the Forums discussion and remove this change unless someone comes up with a use case.